zio-http/shared/src/main/scala/zio/http/Response.scala

```diff
@@ -206,21 +221,83 @@ object Response {
+    fromThrowable(throwable, ErrorResponseConfig.default)
+
   /**
    * Creates a new response for the specified throwable. Note that this method
    * relies on the runtime class of the throwable.
    */
-  def fromThrowable(throwable: Throwable): Response = {
+  def fromThrowable(throwable: Throwable, config: ErrorResponseConfig): Response = {
     throwable match { // TODO: Enhance
-      case _: AccessDeniedException           => error(Status.Forbidden, throwable.getMessage)
-      case _: IllegalAccessException          => error(Status.Forbidden, throwable.getMessage)
-      case _: IllegalAccessError              => error(Status.Forbidden, throwable.getMessage)
-      case _: NotDirectoryException           => error(Status.BadRequest, throwable.getMessage)
-      case _: IllegalArgumentException        => error(Status.BadRequest, throwable.getMessage)
-      case _: java.io.FileNotFoundException   => error(Status.NotFound, throwable.getMessage)
-      case _: java.net.ConnectException       => error(Status.ServiceUnavailable, throwable.getMessage)
-      case _: java.net.SocketTimeoutException => error(Status.GatewayTimeout, throwable.getMessage)
-      case _                                  => error(Status.InternalServerError, throwable.getMessage)
+      case _: AccessDeniedException  => error(Status.Forbidden, throwableToMessage(throwable, Status.Forbidden, config))
+      case _: IllegalAccessException => error(Status.Forbidden, throwableToMessage(throwable, Status.Forbidden, config))
+      case _: IllegalAccessError     => error(Status.Forbidden, throwableToMessage(throwable, Status.Forbidden, config))
+      case _: NotDirectoryException  =>
+        error(Status.BadRequest, throwableToMessage(throwable, Status.BadRequest, config))
+      case _: IllegalArgumentException        =>
+        error(Status.BadRequest, throwableToMessage(throwable, Status.BadRequest, config))
+      case _: java.io.FileNotFoundException   =>
+        error(Status.NotFound, throwableToMessage(throwable, Status.NotFound, config))
+      case _: java.net.ConnectException       =>
+        error(Status.ServiceUnavailable, throwableToMessage(throwable, Status.ServiceUnavailable, config))
+      case _: java.net.SocketTimeoutException =>
+        error(Status.GatewayTimeout, throwableToMessage(throwable, Status.GatewayTimeout, config))
+      case _ => error(Status.InternalServerError, throwableToMessage(throwable, Status.InternalServerError, config))
+    }
+  }
+
+  private def throwableToMessage(throwable: Throwable, status: Status, config: ErrorResponseConfig): Body =
+    if (!config.withErrorBody) Body.empty
+    else {
+      val stackTrace =
+        if (config.withStackTrace && throwable.getStackTrace.nonEmpty)
+          (if (config.maxStackTraceDepth == 0) throwable.getStackTrace
+           else throwable.getStackTrace.take(config.maxStackTraceDepth))
+            .mkString("\n", "\n", "")
+        else ""
```

Comment:
Is `getStackTrace` cheap? Or should we rewrite this to invoke it only once?

Reply:
I'll double check

Reply:
I created a local val. I think we could optimize printing limited stack traces, since `getStackTrace` captures the full stack. We could probably walk the trace only until n. But I think this is okay for now.